node/bindnode/example_test.go

```diff
@@ -80,3 +82,104 @@ func ExamplePrototype_onlySchema() {
 	// Output:
 	// {"Friends":["Sarah","Alex"],"Name":"Michael"}
 }
+
+func ExampleWrap_withSchemaUsingStringjoinStruct() {
+	ts := schema.TypeSystem{}
+	ts.Init()
+	ts.Accumulate(schema.SpawnString("String"))
+	ts.Accumulate(schema.SpawnStruct("FooBarBaz",
+		[]schema.StructField{
+			schema.SpawnStructField("foo", "String", false, false),
+			schema.SpawnStructField("bar", "String", false, false),
+			schema.SpawnStructField("baz", "String", false, false),
+		},
+		schema.SpawnStructRepresentationStringjoin(":"),
+	))
+	schemaType := ts.TypeByName("FooBarBaz")
+
+	type FooBarBaz struct {
+		Foo string
+		Bar string
+		Baz string
+	}
+	fbb := &FooBarBaz{
+		Foo: "x",
+		Bar: "y",
+		Baz: "z",
+	}
+	node := bindnode.Wrap(fbb, schemaType)
+
+	// Take the representation of the node, and serialize it.
+	nodeRepr := node.Representation()
+	var buf bytes.Buffer
+	dagjson.Encode(nodeRepr, &buf)
+
+	// Output how this was serialized, for the example.
+	fmt.Fprintf(os.Stdout, "json: %s\n", buf.Bytes())
+
+	// Now unmarshal that again and print that too, to show it working both ways.
+	np := bindnode.Prototype(&FooBarBaz{}, schemaType)
+	nb := np.Representation().NewBuilder()
+	err := dagjson.Decode(nb, &buf)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("golang: %#v\n", bindnode.Unwrap(nb.Build()))
+
+	// Output:
+	// json: "x:y:z"
+	// golang: &bindnode_test.FooBarBaz{Foo:"x", Bar:"y", Baz:"z"}
+}
+
+func ExampleWrap_withSchemaUsingStringprefixUnion() {
+	ts := schema.TypeSystem{}
+	ts.Init()
+	ts.Accumulate(schema.SpawnString("Foo"))
+	ts.Accumulate(schema.SpawnString("Bar"))
+	ts.Accumulate(schema.SpawnUnion("FooOrBar",
+		[]schema.TypeName{
+			"Foo",
+			"Bar",
+		},
+		schema.SpawnUnionRepresentationStringprefix(
+			":", // n.b. this API will change soon; a schema-schema iteration has removed the distinct joiner string.
+			map[string]schema.TypeName{
+				"foo": "Foo",
+				"bar": "Bar",
+			},
+		),
+	))
+	schemaType := ts.TypeByName("FooOrBar")
+
+	// The golang structures for unions don't look as simple as one might like.  Golang has no native sum types, so we do something interesting here.
+	type FooOrBar struct {
+		Index int
+		Value interface{}
+	}
```

Comment:
is this really the best we can do?

Reply:
This seems like a reasonable solution to an awkward Go limitation. Other options would probably involve nested structs which starts to detract from the concept of a union, or is there some other option here?
Maybe unions should have some optionality to back them by different strategies depending on user preference.

Reply:
I definitely intend to revisit this. "integer plus empty interface" just was the simplest implementation that let me push forward with the implementation - and it's also what codegen uses.

It's possible that we could support just an interface value, and detect what type is currently stored via the Go type. Though that would require some extra logic, so I didn't want to implement that straight away.

Another option would be to make it arbitrary logic, like:

```
type Union interface {
    Set(idx int, value interface{})
    Get() (idx int, value interface{})
}
```

I think it makes sense to wait seven months for generics to land to revisit this, because generics would potentially offer us a far better option:

```
type FooOrBar interface {
    type Foo, Bar
}
```

Generics would also open more doors in terms of expressing custom map implementations, optional types, nullable types, etc - all without the `interface{}` shenanigans.

Reply:
On second thought, the current way of dealing with unions is just plain bad. See https://github.com/ipld/go-ipld-prime/issues/210. I'm marking this as resolved because they'll need to be reworked anyway.